File: src/main/java/com/freya02/botcommands/api/components/ComponentInteractionFilter.java
package com.freya02.botcommands.api.components;

import com.freya02.botcommands.api.commands.application.ApplicationCommandFilter;
import com.freya02.botcommands.api.core.CooldownService;
import com.freya02.botcommands.api.core.config.BServiceConfigBuilder;
import com.freya02.botcommands.api.core.service.annotations.BService;
import com.freya02.botcommands.api.core.service.annotations.InterfacedService;
import net.dv8tion.jda.api.events.interaction.component.GenericComponentInteractionCreateEvent;
import org.jetbrains.annotations.NotNull;

/**
 * Filters component interactions (such as buttons and select menus),
 * any filter that returns {@code false} prevents the interaction from executing.
 *
 * <p>Filters are tested right before the component gets executed (i.e., after the component constraints were checked).
 *
 * <p><b>Note:</b> Your filter still has to acknowledge the interaction in case it rejects it.
 *
 * <p><b>Usage</b>: Register your instance as a service with {@link BService}
 * or {@link BServiceConfigBuilder#getServiceAnnotations() any annotation that enables your class for dependency injection}.
 *
 * @see InterfacedService @InterfacedService
 *
 * @see #isAccepted(GenericComponentInteractionCreateEvent)
 * @see CooldownService
 */
@InterfacedService(acceptMultiple = true)
public interface ComponentInteractionFilter {
	/**
	 * Returns whether the component interaction should be accepted or not.
	 *
	 * <p><b>Note:</b> Your filter still has to acknowledge the interaction in case it rejects it.
	 *
	 * @return {@code true} if the component interaction can run, {@code false} otherwise
	 *
	 * @see ApplicationCommandFilter
	 */
	boolean isAccepted(@NotNull GenericComponentInteractionCreateEvent event);
}

File: src/main/kotlin/com/freya02/botcommands/api/components/ComponentInteractionFilter.kt
package com.freya02.botcommands.api.components

import com.freya02.botcommands.api.core.config.BServiceConfigBuilder
import com.freya02.botcommands.api.core.service.annotations.BService
import com.freya02.botcommands.api.core.service.annotations.InterfacedService
import com.freya02.botcommands.api.core.utils.simpleNestedName
import net.dv8tion.jda.api.events.interaction.component.GenericComponentInteractionCreateEvent

/**
 * Filters component interactions (such as buttons and select menus),
 * any filter that returns `false` prevents the interaction from executing.
 *
 * Filters are tested right before the component gets executed (i.e., after the constraints/rate limits were checked).
 *
 * **Note:** Your filter still has to acknowledge the interaction in case it rejects it.
 *
 * **Usage**: Register your instance as a service with [BService]
 * or [any annotation that enables your class for dependency injection][BServiceConfigBuilder.serviceAnnotations].
 *
 * **Example** - Rejecting component interactions from non-owners:
 * ```kt
 * @BService
 * class MyComponentFilter(private val config: BConfig) : ComponentInteractionFilter {
 *     override suspend fun isAcceptedSuspend(event: GenericComponentInteractionCreateEvent): Boolean {
 *         if (event.user.idLong !in config.ownerIds) {
 *             event.reply_("Only owners are allowed to use components", ephemeral = true).queue()
 *             return false
 *         }
 *         return true
 *     }
 * }
 * ```
 *
 * <Hr>
 *
 * ```java
 * @BService
 * public class MyComponentCommandFilter implements ApplicationCommandFilter {
 *     private final BConfig config;
 *
 *     public MyComponentCommandFilter(BConfig config) {
 *         this.config = config;
 *     }
 *
 *     @Override
 *     public boolean isAccepted(@NotNull GenericCommandInteractionEvent event, @NotNull ApplicationCommandInfo commandInfo) {
 *         if (!config.isOwner(event.getUser().getIdLong())) {
 *             event.reply("Only owners are allowed to use components").setEphemeral(true).queue();
 *             return false;
 *         }
 *         return true;
 *     }
 * }
 * ```
 *
 * @see InterfacedService @InterfacedService
 *
 * @see isAccepted
 */
@InterfacedService(acceptMultiple = true)
interface ComponentInteractionFilter {
    /**
     * Returns whether the component interaction should be accepted or not.
     *
     * **Note:** Your filter still has to acknowledge the interaction in case it rejects it.
     *
     * @return `true` if the component interaction can run, `false` otherwise
     *
     * @see ComponentInteractionFilter
     */
    @JvmSynthetic
    suspend fun isAcceptedSuspend(event: GenericComponentInteractionCreateEvent): Boolean =
        isAccepted(event)

    /**
     * Returns whether the component interaction should be accepted or not.
     *
     * **Note:** Your filter still has to acknowledge the interaction in case it rejects it.
     *
     * @return `true` if the component interaction can run, `false` otherwise
     *
     * @see ComponentInteractionFilter
     */
    fun isAccepted(event: GenericComponentInteractionCreateEvent): Boolean =
        throw NotImplementedError("${this.javaClass.simpleNestedName} must implement the 'isAccepted' or 'isAcceptedSuspend' method")
}
